File: research/experiments/sweep/runs/opt-large-go-sonnet45/line_search.go
package optimization

import "math"

// LineSearchResult contains the result of a line search
type LineSearchResult struct {
	Alpha         float64
	FNew          float64
	GNew          []float64 // nil if not computed
	FunctionCalls int
	GradientCalls int
	Success       bool
}

// BacktrackingLineSearch performs a simple backtracking line search (Armijo condition)
func BacktrackingLineSearch(
	f func([]float64) float64,
	x, d []float64,
	fx float64,
	gx []float64,
	c1 float64,
	rho float64,
	initialAlpha float64,
	maxIter int,
) LineSearchResult {
	// Set defaults
	if c1 == 0 {
		c1 = 1e-4
	}
	if rho == 0 {
		rho = 0.5
	}
	if initialAlpha == 0 {
		initialAlpha = 1.0
	}
	if maxIter == 0 {
		maxIter = 20
	}

	alpha := initialAlpha
	gxDotD := Dot(gx, d)
	functionCalls := 0

	for i := 0; i < maxIter; i++ {
		xNew := AddScaled(x, d, alpha)
		fNew := f(xNew)
		functionCalls++

		// Check Armijo condition
		if fNew <= fx+c1*alpha*gxDotD {
			return LineSearchResult{
				Alpha:         alpha,
				FNew:          fNew,
				GNew:          nil,
				FunctionCalls: functionCalls,
				GradientCalls: 0,
				Success:       true,
			}
		}

		alpha *= rho
	}

	// Failed to find acceptable step
	return LineSearchResult{
		Alpha:         alpha,
		FNew:          f(AddScaled(x, d, alpha)),
		GNew:          nil,
		FunctionCalls: functionCalls + 1,
		GradientCalls: 0,
		Success:       false,
	}
}

// WolfeLineSearch performs a line search satisfying the strong Wolfe conditions
func WolfeLineSearch(
	f func([]float64) float64,
	grad func([]float64) []float64,
	x, d []float64,
	fx float64,
	gx []float64,
	c1 float64,
	c2 float64,
	alphaMax float64,
	maxIter int,
) LineSearchResult {
	// Set defaults
	if c1 == 0 {
		c1 = 1e-4
	}
	if c2 == 0 {
		c2 = 0.9
	}
	if alphaMax == 0 {
		alphaMax = 1e6
	}
	if maxIter == 0 {
		maxIter = 25
	}

	functionCalls := 0
	gradientCalls := 0
	gxDotD := Dot(gx, d)

	alpha1 := 1.0
	phi0 := fx
	dphi0 := gxDotD

	alphaPrev := 0.0
	phiPrev := phi0

	for i := 0; i < maxIter; i++ {
		xNew := AddScaled(x, d, alpha1)
		phi1 := f(xNew)
		functionCalls++

		// Check if we violate Armijo or if function increased from previous
		if (phi1 > phi0+c1*alpha1*dphi0) || (i > 0 && phi1 >= phiPrev) {
			result := zoom(f, grad, x, d, alphaPrev, alpha1, phi0, dphi0, phiPrev, phi1, c1, c2, &functionCalls, &gradientCalls)
			return result
		}

		gNew := grad(xNew)
		gradientCalls++
		dphi1 := Dot(gNew, d)

		// Check curvature condition
		if math.Abs(dphi1) <= -c2*dphi0 {
			return LineSearchResult{
				Alpha:         alpha1,
				FNew:          phi1,
				GNew:          gNew,
				FunctionCalls: functionCalls,
				GradientCalls: gradientCalls,
				Success:       true,
			}
		}

		// Check if slope is positive (we've gone past minimum)
		if dphi1 >= 0 {
			result := zoom(f, grad, x, d, alpha1, alphaPrev, phi0, dphi0, phi1, phiPrev, c1, c2, &functionCalls, &gradientCalls)
			return result
		}

		// Expand bracket
		alphaPrev = alpha1
		phiPrev = phi1
		alpha1 = math.Min(alpha1*2.0, alphaMax)
	}

	// Failed to find acceptable step
	xNew := AddScaled(x, d, alpha1)
	fNew := f(xNew)
	functionCalls++

	return LineSearchResult{
		Alpha:         alpha1,
		FNew:          fNew,
		GNew:          nil,
		FunctionCalls: functionCalls,
		GradientCalls: gradientCalls,
		Success:       false,
	}
}

// zoom is the internal function for narrowing the bracket in Wolfe line search
func zoom(
	f func([]float64) float64,
	grad func([]float64) []float64,
	x, d []float64,
	alphaLo, alphaHi float64,
	phi0, dphi0 float64,
	phiLo, phiHi float64,
	c1, c2 float64,
	functionCalls *int,
	gradientCalls *int,
) LineSearchResult {
	const maxZoomIter = 20

	for i := 0; i < maxZoomIter; i++ {
		// Bisection (simple approach; cubic interpolation would be more sophisticated)
		alpha := (alphaLo + alphaHi) / 2.0

		xNew := AddScaled(x, d, alpha)
		phi := f(xNew)
		*functionCalls++

		if phi > phi0+c1*alpha*dphi0 || phi >= phiLo {
			alphaHi = alpha
			phiHi = phi
		} else {
			gNew := grad(xNew)
			*gradientCalls++
			dphi := Dot(gNew, d)

			if math.Abs(dphi) <= -c2*dphi0 {
				return LineSearchResult{
					Alpha:         alpha,
					FNew:          phi,
					GNew:          gNew,
					FunctionCalls: *functionCalls,
					GradientCalls: *gradientCalls,
					Success:       true,
				}
			}

			if dphi*(alphaHi-alphaLo) >= 0 {
				alphaHi = alphaLo
				phiHi = phiLo
			}

			alphaLo = alpha
			phiLo = phi
		}
	}

	// Failed to zoom in on acceptable step
	alpha := (alphaLo + alphaHi) / 2.0
	xNew := AddScaled(x, d, alpha)
	fNew := f(xNew)
	*functionCalls++

	return LineSearchResult{
		Alpha:         alpha,
		FNew:          fNew,
		GNew:          nil,
		FunctionCalls: *functionCalls,
		GradientCalls: *gradientCalls,
		Success:       false,
	}
}

File: research/experiments/sweep/runs/opt-full-csharp-sonnet45/src/LineSearch.cs
namespace Optimization;

public class LineSearchResult
{
    public double Alpha { get; set; }
    public double FNew { get; set; }
    public double[]? GNew { get; set; }
    public int FunctionCalls { get; set; }
    public int GradientCalls { get; set; }
    public bool Success { get; set; }
}

/// <summary>
/// Backtracking (Armijo) and Strong Wolfe line search.
/// </summary>
public static class LineSearch
{
    public static LineSearchResult BacktrackingLineSearch(
        Func<double[], double> f,
        double[] x, double[] d, double fx, double[] gx,
        double initialAlpha = 1.0, double c1 = 1e-4, double rho = 0.5, int maxIter = 20)
    {
        double dirDeriv = VecOps.Dot(gx, d);
        double alpha = initialAlpha;
        int functionCalls = 0;

        for (int i = 0; i < maxIter; i++)
        {
            double[] xNew = VecOps.AddScaled(x, d, alpha);
            double fNew = f(xNew);
            functionCalls++;

            if (fNew <= fx + c1 * alpha * dirDeriv)
            {
                return new LineSearchResult
                {
                    Alpha = alpha, FNew = fNew, GNew = null,
                    FunctionCalls = functionCalls, GradientCalls = 0, Success = true
                };
            }
            alpha *= rho;
        }

        double[] xFinal = VecOps.AddScaled(x, d, alpha);
        double fFinal = f(xFinal);
        functionCalls++;
        return new LineSearchResult
        {
            Alpha = alpha, FNew = fFinal, GNew = null,
            FunctionCalls = functionCalls, GradientCalls = 0, Success = false
        };
    }

    public static LineSearchResult WolfeLineSearch(
        Func<double[], double> f, Func<double[], double[]> grad,
        double[] x, double[] d, double fx, double[] gx,
        double c1 = 1e-4, double c2 = 0.9, double alphaMax = 1e6, int maxIter = 25)
    {
        double dphi0 = VecOps.Dot(gx, d);
        int functionCalls = 0;
        int gradientCalls = 0;

        double alphaPrev = 0;
        double phiPrev = fx;
        double alpha = 1.0;

        for (int i = 0; i < maxIter; i++)
        {
            double[] xNew = VecOps.AddScaled(x, d, alpha);
            double phiAlpha = f(xNew);
            functionCalls++;

            // Armijo violation or not sufficient decrease compared to previous
            if (phiAlpha > fx + c1 * alpha * dphi0 || (i > 0 && phiAlpha >= phiPrev))
            {
                return Zoom(f, grad, x, d, fx, dphi0, alphaPrev, alpha, phiPrev, phiAlpha,
                    c1, c2, ref functionCalls, ref gradientCalls);
            }

            double[] gNew = grad(xNew);
            gradientCalls++;
            double dphiAlpha = VecOps.Dot(gNew, d);

            // Strong Wolfe curvature condition
            if (Math.Abs(dphiAlpha) <= c2 * Math.Abs(dphi0))
            {
                return new LineSearchResult
                {
                    Alpha = alpha, FNew = phiAlpha, GNew = gNew,
                    FunctionCalls = functionCalls, GradientCalls = gradientCalls, Success = true
                };
            }

            if (dphiAlpha >= 0)
            {
                return Zoom(f, grad, x, d, fx, dphi0, alpha, alphaPrev, phiAlpha, phiPrev,
                    c1, c2, ref functionCalls, ref gradientCalls);
            }

            alphaPrev = alpha;
            phiPrev = phiAlpha;
            alpha = Math.Min(alpha * 2, alphaMax);
        }

        // Failed
        double[] xF = VecOps.AddScaled(x, d, alpha);
        double fF = f(xF);
        functionCalls++;
        double[] gF = grad(xF);
        gradientCalls++;
        return new LineSearchResult
        {
            Alpha = alpha, FNew = fF, GNew = gF,
            FunctionCalls = functionCalls, GradientCalls = gradientCalls, Success = false
        };
    }

    private static LineSearchResult Zoom(
        Func<double[], double> f, Func<double[], double[]> grad,
        double[] x, double[] d, double fx, double dphi0,
        double alphaLo, double alphaHi, double phiLo, double phiHi,
        double c1, double c2, ref int functionCalls, ref int gradientCalls)
    {
        for (int j = 0; j < 20; j++)
        {
            double alpha = (alphaLo + alphaHi) / 2;
            double[] xNew = VecOps.AddScaled(x, d, alpha);
            double phiAlpha = f(xNew);
            functionCalls++;

            if (phiAlpha > fx + c1 * alpha * dphi0 || phiAlpha >= phiLo)
            {
                alphaHi = alpha;
                phiHi = phiAlpha;
            }
            else
            {
                double[] gNew = grad(xNew);
                gradientCalls++;
                double dphiAlpha = VecOps.Dot(gNew, d);

                if (Math.Abs(dphiAlpha) <= c2 * Math.Abs(dphi0))
                {
                    return new LineSearchResult
                    {
                        Alpha = alpha, FNew = phiAlpha, GNew = gNew,
                        FunctionCalls = functionCalls, GradientCalls = gradientCalls, Success = true
                    };
                }

                if (dphiAlpha * (alphaHi - alphaLo) >= 0)
                {
                    alphaHi = alphaLo;
                    phiHi = phiLo;
                }

                alphaLo = alpha;
                phiLo = phiAlpha;
            }
        }

        double[] xFinal = VecOps.AddScaled(x, d, alphaLo);
        double fFinal = f(xFinal);
        functionCalls++;
        double[] gFinal = grad(xFinal);
        gradientCalls++;
        return new LineSearchResult
        {
            Alpha = alphaLo, FNew = fFinal, GNew = gFinal,
            FunctionCalls = functionCalls, GradientCalls = gradientCalls, Success = false
        };
    }
}
